Mathlib/Data/Set/Function.lean

```diff
@@ -0,0 +1,1548 @@
+/-
+Copyright (c) 2014 Jeremy Avigad. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jeremy Avigad, Andrew Zipperer, Haitao Zhang, Minchao Wu, Yury Kudryashov
+
+! This file was ported from Lean 3 source module data.set.function
+! leanprover-community/mathlib commit 198161d833f2c01498c39c266b0b3dbe2c7a8c07
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Data.Set.Prod
+import Mathlib.Logic.Function.Conjugate
+
+/-!
+# Functions over sets
+
+## Main definitions
+
+### Predicate
+
+* `Set.EqOn f₁ f₂ s` : functions `f₁` and `f₂` are equal at every point of `s`;
+* `Set.MapsTo f s t` : `f` sends every point of `s` to a point of `t`;
+* `Set.InjOn f s` : restriction of `f` to `s` is injective;
+* `Set.SurjOn f s t` : every point in `s` has a preimage in `s`;
+* `Set.BijOn f s t` : `f` is a bijection between `s` and `t`;
+* `Set.LeftInvOn f' f s` : for every `x ∈ s` we have `f' (f x) = x`;
+* `Set.RightInvOn f' f t` : for every `y ∈ t` we have `f (f' y) = y`;
+* `Set.InvOn f' f s t` : `f'` is a two-side inverse of `f` on `s` and `t`, i.e.
+  we have `Set.LeftInvOn f' f s` and `Set.RightInvOn f' f t`.
+
+### Functions
+
+* `Set.restrict f s` : restrict the domain of `f` to the set `s`;
+* `Set.cod_restrict f s h` : given `h : ∀ x, f x ∈ s`, restrict the codomain of `f` to the set `s`;
+* `Set.MapsTo.restrict f s t h`: given `h : MapsTo f s t`, restrict the domain of `f` to `s`
+  and the codomain to `t`.
+-/
+
+variable {α β γ ι : Type _} {π : α → Type _}
+
+open Function
+
+namespace Set
+
+/-! ### Restrict -/
+
+
+/-- Restrict domain of a function `f` to a set `s`. Same as `subtype.restrict` but this version
+takes an argument `↥s` instead of `subtype s`. -/
```

Comment:
```suggestion
/-- Restrict domain of a function `f` to a set `s`. Same as `Subtype.restrict` but this version
takes an argument `↥s` instead of `Subtype s`. -/
```